README
= Darkfish Rdoc Generator

This is a complete replacement for the default HTML generator for Rdoc 2. 

I created it with a few goals in mind:

* Replace the four-frame layout with three different indexes with a single-page, 
  class/module-focused view.

* I helped write what became the default look and feel for Rdoc, but it's 
  starting to show its age. I wanted to update it with new CSS and Javascript 
  for modern browsers.

* Add some of the nifty searching and index-on-the-side stuff I liked from the
  modified Allison template (http://www.superalloy.nl/blog/?p=7) Eloy 
  Duran came up with.

* Implemented using separate ERB template files instead of Rdoc's 
  constants-in-a-module style templating.

I think I mostly succeeded, and it's nice enough that I wanted to use it for all
my documentation, and I thought you might want to, too. Or at least be able to
tweak it to generate your own spiffy docs.

It looks like this:

http://deveiate.org/darkfish_preview.png

== Getting Started

If you haven't installed it yet, you can install it via RubyGems:

 $ sudo gem install darkfish-rdoc

Or download the source and do:
	
 $ sudo rake install

== Usage

If you're using RDoc 2.3 or later, the Darkfish formatter will be automatically discovered when RDoc
is loaded, so you only have to specify that you want the 'darkfish' formatter:

 $ rdoc -w 4 -SHN -f darkfish -m README README lib

It's equally easy to specify from a Rakefile in an RDocTask:

  ### Task: rdoc
  require 'rake/rdoctask'
  gem 'darkfish-rdoc' # Omit this if it's installed as a regular library

  Rake::RDocTask.new do |rdoc|
    rdoc.title    = "My Awesome SnoCone Library"
  
    rdoc.options += [
        '-w', '4',
        '-SHN',
        '-f', 'darkfish', # This bit
        '-m', 'README',
      ]
    
    rdoc.rdoc_files.include 'README'
    rdoc.rdoc_files.include 'lib/snocone.rb'
  end
  
  
That's it.

== Feedback

I'd be happy to have feedback via email (ged at FaerieMUD dot org)
or a ticket on the project's Trac:

  http://deveiate.org/projects/Darkfish-Rdoc/

== License

See the LICENSE file included with the distribution for copyright and licensing details.



ChangeLog
-- Wed, 07 Jan 2009 02:52:03 -0000 by deveiant (r53) -----
   Added:   lib/rdoc/generator/template/darkfish/images/find.png (new)
            lib/rdoc/generator/template/darkfish/images/bullet_toggle_minus.png (new)
            lib/rdoc/generator/template/darkfish/images/bullet_toggle_plus.png (new)
   Changed: lib/rdoc/generator/template/darkfish/classpage.rhtml
            lib/rdoc/generator/template/darkfish/js/quicksearch.js
            ChangeLog
            lib/rdoc/generator/template/darkfish/rdoc.css
            lib/rdoc/generator/template/darkfish/js/darkfish.js

   Some changes suggested by Eric Hodel (refs #5)
   * Moved quicksearch up to the top of the class index; still needs
   CSS fixing.
   * Fixed mistakenly-applied roundbox to DLs in simple file pages


-- Wed, 07 Jan 2009 02:08:11 -0000 by deveiant (r52) -----
   Changed: lib/rdoc/generator/darkfish.rb
            lib/rdoc/generator/template/darkfish/classpage.rhtml
            lib/rdoc/generator/template/darkfish/index.rhtml
            lib/rdoc/generator/template/darkfish/rdoc.css
            lib/rdoc/generator/template/darkfish/filepage.rhtml

   Merged patches from Eric Hodel (drbrain at segment7.net). Refs #5.


-- Sat, 20 Dec 2008 21:04:57 -0000 by deveiant (r51) -----
   Added:   lib/rdoc/generator/template/darkfish/images/bug.png (new)
   Changed: lib/rdoc/generator/darkfish.rb
            lib/rdoc/generator/template/darkfish/classpage.rhtml
            lib/rdoc/generator/template/darkfish/index.rhtml
            ChangeLog
            lib/rdoc/generator/template/darkfish/rdoc.css
            lib/rdoc/generator/template/darkfish/filepage.rhtml (and 3 other/s)

    * Make the javascript headers all load the same thing.
    * Moved the dom-ready handler into darkfish.js instead of a script section in the head.
    * Insert RDoc debugging output when $DEBUG is true.
    * Change the email addresses in darkfish.rb from <address>-style to (address) as a workaround so
    RDoc output is valid.
    * Updated the instructions for use in the README.


-- Thu, 18 Dec 2008 16:39:30 -0000 by deveiant (r50) -----
   Changed: lib/rdoc/generator/darkfish.rb
            LICENSE
            ChangeLog
            lib/rdoc/generator/template/darkfish/rdoc.css
            README

    * CSS fix: Use the same formatting as the index page for content in file pages.
    * Merged license statements.
    * Added Mahlon to the contributors section.


-- Thu, 18 Dec 2008 00:59:26 -0000 by deveiant (r49) -----
   Changed: lib/rdoc/generator/darkfish.rb

   Bumping release version to 1.1.6.


-- Thu, 18 Dec 2008 00:58:30 -0000 by deveiant (r48) -----
   Added:   lib/rdoc/generator/template/darkfish/images/page_white_text.png (new)
   Changed: spec/rdoc/generator/darkfish_spec.rb
            lib/rdoc/generator/darkfish.rb
            lib/rdoc/generator/template/darkfish/classpage.rhtml
            LICENSE
            Rakefile
            lib/rdoc/generator/template/darkfish/index.rhtml (and 6 other/s)

    * Merged patches from Eric Hodel for Rdoc 2.3 changes + improvements (closes #4)
    - Symbol keys -> strings
    - Added a Files list to the index page
    - Support Rubygems auto-discovery
    * Reworked the file page template so it renders differently for 'SimpleParser' pages.
    * Cleaned up the DL-related classes so constants don't look so weird.
    * Removed the box around H1s. -ETOOMANYBOXES


-- Wed, 17 Dec 2008 22:26:58 -0000 by deveiant (r47) -----
   Added:   /branches/rdoc-2.3-changes
            -> renamed from /trunk@r46

   Making a branch for RDoc 2.3 changes so I don't break the build


-- Thu, 04 Dec 2008 22:31:39 -0000 by deveiant (r46) -----
   Changed: /trunk/Rakefile
            /trunk/project.yml

   Updated build system.


-- Thu, 04 Dec 2008 19:12:13 -0000 by deveiant (r45) -----
   Changed: /trunk/spec/rdoc/generator/darkfish_spec.rb

   Fixed broken expectation setup.


-- Thu, 04 Dec 2008 19:07:12 -0000 by deveiant (r44) -----
   Changed: /trunk
            /trunk/lib/rdoc/generator/template/darkfish/rdoc.css

   Cleaned up display of constant and attribute lists.


-- Thu, 04 Dec 2008 06:50:02 -0000 by deveiant (r43) -----
   Added:   /trunk/lib/rdoc/discover.rb
            -> renamed from /trunk/lib/darkfish-rdoc.rb@r42
   Changed: /trunk/Rakefile
            /trunk/project.yml
   Deleted: /trunk/lib/darkfish-rdoc.rb

    * Integrated patch from Eric Hodel (drbrain at segment7.net) for RDoc 2.2.2 autodiscovery. Thanks
    Eric!
    * Updated build system


-- Mon, 06 Oct 2008 21:33:22 -0000 by deveiant (r42) -----
   Changed: /trunk/Rakefile

   Updated build system.


-- Fri, 26 Sep 2008 06:08:54 -0000 by deveiant (r40) -----
   Changed: /trunk/ChangeLog
            /trunk/lib/rdoc/generator/template/darkfish/rdoc.css

   Fixed the margin/border on constants and attributes


-- Thu, 25 Sep 2008 16:50:52 -0000 by deveiant (r38) -----
   Changed: /trunk/lib/rdoc/generator/template/darkfish/js/jquery.js
            /trunk/lib/rdoc/generator/template/darkfish/js/quicksearch.js
            /trunk/Rakefile.local

    * Updated jquery to 1.2.6
    * Focus the results of a quicksearch when it is off screen
    * Fixed the 'install' Rake task -- it now installs the templates and resources in 
    addition to the ruby code.


-- Mon, 22 Sep 2008 15:57:32 -0000 by deveiant (r36) -----
   Changed: /trunk/Rakefile
            /trunk/ChangeLog

   Updated build system to add the workaround for broken developmental dependencies under RubyGems 1.2.0.


-- Mon, 22 Sep 2008 15:14:43 -0000 by deveiant (r35) -----
   Changed: /trunk/spec/rdoc/generator/darkfish_spec.rb
            /trunk/lib/rdoc/generator/darkfish.rb

   Updated to work under RDoc 2.2.0 (fixes #2).


-- Tue, 16 Sep 2008 14:22:21 -0000 by deveiant (r34) -----
   Changed: /trunk/Rakefile
            /trunk/project.yml

   Updated the build system.


-- Tue, 16 Sep 2008 14:21:23 -0000 by deveiant (r33) -----
   Changed: /trunk/lib/rdoc/generator/darkfish.rb

lib/rdoc/generator/darkfish.rb
#!ruby

require 'rubygems'
gem 'rdoc', '>= 2.3'

require 'pp'
require 'pathname'
require 'fileutils'
require 'erb'
require 'yaml'

require 'rdoc/rdoc'
require 'rdoc/generator/xml'
require 'rdoc/generator/html'

#
#  Darkfish RDoc HTML Generator
#  
#  $Id$
#
#  == Author/s
#  * Michael Granger (ged@FaerieMUD.org)
#  
#  == Contributors
#  * Mahlon E. Smith (mahlon@martini.nu)
#  * Eric Hodel (drbrain@segment7.net)
#  
#  == License
#  
#  :include: LICENSE
#  
class RDoc::Generator::Darkfish < RDoc::Generator::XML

	RDoc::RDoc.add_generator( self )

	include ERB::Util

	# Subversion rev
	SVNRev = %$Rev$
	
	# Subversion ID
	SVNId = %$Id$

	# Path to this file's parent directory. Used to find templates and other
	# resources.
	GENERATOR_DIR = Pathname.new( __FILE__ ).expand_path.dirname

	# Release Version
	VERSION = '1.1.6'

	# Directory where generated classes live relative to the root
	CLASS_DIR = nil

	# Directory where generated files live relative to the root
	FILE_DIR = nil


	#################################################################
	###	C L A S S   M E T H O D S
	#################################################################

	### Standard generator factory method
	def self::for( options )
		new( options )
	end


	#################################################################
	###	I N S T A N C E   M E T H O D S
	#################################################################

	### Initialize a few instance variables before we start
	def initialize( options )
		@template = nil
		@template_dir = GENERATOR_DIR + 'template/darkfish'
		
		@files      = []
		@classes    = []
		@hyperlinks = {}

		@basedir = Pathname.pwd.expand_path

		options.inline_source = true
		options.diagram = false

		super
	end
	
	
	######
	public
	######

	# The output directory
	attr_reader :outputdir
	
	
	### Output progress information if debugging is enabled
	def debug_msg( *msg )
		return unless $DEBUG
		$stderr.puts( *msg )
	end
	
	
	### Create the directories the generated docs will live in if
	### they don't already exist.
	def gen_sub_directories
		@outputdir.mkpath
	end
	

	### Copy over the stylesheet into the appropriate place in the
	### output directory.
	def write_style_sheet
		debug_msg "Copying over static files"
		staticfiles = %w[rdoc.css js images]
		staticfiles.each do |path|
			FileUtils.cp_r( @template_dir + path, '.', :verbose => $DEBUG, :noop => $dryrun )
		end
	end
	
	

	### Build the initial indices and output objects
	### based on an array of TopLevel objects containing
	### the extracted information. 
	def generate( toplevels )
		@outputdir = Pathname.new( @options.op_dir ).expand_path( @basedir )
		if RDoc::Generator::Context.respond_to?( :build_indicies)
	    	@files, @classes = RDoc::Generator::Context.build_indicies( toplevels, @options )
		else
	    	@files, @classes = RDoc::Generator::Context.build_indices( toplevels, @options )
		end

		# Now actually write the output
		generate_xhtml( @options, @files, @classes )

	rescue StandardError => err
		debug_msg "%s: %s\n  %s" % [ err.class.name, err.message, err.backtrace.join("\n  ") ]
		raise
	end


	### No-opped
	def load_html_template # :nodoc:
	end


	### Generate output
	def generate_xhtml( options, files, classes )
		files = gen_into( @files )
		classes = gen_into( @classes )

		# Make a hash of class info keyed by class name
		classes_by_classname = classes.inject({}) {|hash, classinfo|
			hash[ classinfo[:full_name] ] = classinfo
			hash[ classinfo[:full_name] ][:outfile] =
				classinfo[:full_name].gsub( /::/, '/' ) + '.html'
			hash
		}

		# Make a hash of file info keyed by path
		files_by_path = files.inject({}) {|hash, fileinfo|
			hash[ fileinfo[:full_path] ] = fileinfo
			hash[ fileinfo[:full_path] ][:outfile] = fileinfo[:full_path] + '.html'
			hash
		}

		self.write_style_sheet
		self.generate_index( options, files_by_path, classes_by_classname )
		self.generate_class_files( options, files_by_path, classes_by_classname )
		self.generate_file_files( options, files_by_path, classes_by_classname )
	end



	#########
	protected
	#########

	### Return a list of the documented modules sorted by salience first, then by name.
	def get_sorted_module_list( classes )
		nscounts = classes.keys.inject({}) do |counthash, name|
			toplevel = name.gsub( /::.*/, '' )
			counthash[toplevel] ||= 0
			counthash[toplevel] += 1
			
			counthash
		end

		# Sort based on how often the toplevel namespace occurs, and then on the name 
		# of the module -- this works for projects that put their stuff into a 
		# namespace, of course, but doesn't hurt if they don't.
		return classes.keys.sort_by do |name| 
			toplevel = name.gsub( /::.*/, '' )
			[
				nscounts[ toplevel ] * -1,
				name
			]
		end
	end
	
	
	### Generate an index page which lists all the classes which
	### are documented.
	def generate_index( options, files, classes )
		debug_msg "Rendering the index page..."

		templatefile = @template_dir + 'index.rhtml'
		template_src = templatefile.read
		template = ERB.new( template_src, nil, '<>' )
		template.filename = templatefile.to_s
		context = binding()

		modsort = self.get_sorted_module_list( classes )
		output = nil
		begin
			output = template.result( context )
		rescue NoMethodError => err
			raise "Error while evaluating %s: %s (at %p)" % [
				templatefile,
				err.message,
				eval( "_erbout[-50,50]", context )
			]
		end

		outfile = @basedir + @options.op_dir + 'index.html'
		unless $dryrun
			debug_msg "Outputting to %s" % [outfile.expand_path]
			outfile.open( 'w', 0644 ) do |fh|
				fh.print( output )
			end
		else
			debug_msg "Would have output to %s" % [outfile.expand_path]
		end
	end



	### Generate a documentation file for each class present in the
	### given hash of +classes+.
	def generate_class_files( options, files, classes )
		debug_msg "Generating class documentation in #@outputdir"
		templatefile = @template_dir + 'classpage.rhtml'
		outputdir = @outputdir

		modsort = self.get_sorted_module_list( classes )

		classes.sort_by {|k,v| k }.each do |classname, classinfo|
			debug_msg "  working on %s (%s)" % [ classname, classinfo[:outfile] ]
			outfile    = outputdir + classinfo[:outfile]
			rel_prefix = outputdir.relative_path_from( outfile.dirname )
			svninfo    = self.get_svninfo( classinfo )

			debug_msg "  rendering #{outfile}"
			self.render_template( templatefile, binding(), outfile )
		end
	end


	### Generate a documentation file for each file present in the
	### given hash of +files+.
	def generate_file_files( options, files, classes )
		debug_msg "Generating file documentation in #@outputdir"
		templatefile = @template_dir + 'filepage.rhtml'

		modsort = self.get_sorted_module_list( classes )

		files.sort_by {|k,v| k }.each do |path, fileinfo|
			outfile     = @outputdir + fileinfo[:outfile]
			debug_msg "  working on %s (%s)" % [ path, outfile ]
			rel_prefix  = @outputdir.relative_path_from( outfile.dirname )
			context     = binding()

			debug_msg "  rendering #{outfile}"
			self.render_template( templatefile, binding(), outfile )
		end
	end


	### Return a string describing the amount of time in the given number of
	### seconds in terms a human can understand easily.
	def time_delta_string( seconds )
		return 'less than a minute' if seconds < 1.minute 
		return (seconds / 1.minute).to_s + ' minute' + (seconds/60 == 1 ? '' : 's') if seconds < 50.minutes
		return 'about one hour' if seconds < 90.minutes
		return (seconds / 1.hour).to_s + ' hours' if seconds < 18.hours
		return 'one day' if seconds < 1.day
		return 'about one day' if seconds < 2.days
		return (seconds / 1.day).to_s + ' days' if seconds < 1.week
		return 'about one week' if seconds < 2.week
		return (seconds / 1.week).to_s + ' weeks' if seconds < 3.months
		return (seconds / 1.month).to_s + ' months' if seconds < 1.year
		return (seconds / 1.year).to_s + ' years'
	end


	# %q$Id$"
	SVNID_PATTERN = /
		\$Id:\s 
			(\S+)\s					# filename
			(\d+)\s					# rev
			(\d{4}-\d{2}-\d{2})\s	# Date (YYYY-MM-DD)
			(\d{2}:\d{2}:\d{2}Z)\s	# Time (HH:MM:SSZ)
			(\w+)\s				 	# committer
		\$$
	/x

	### Try to extract Subversion information out of the first constant whose value looks like
	### a subversion Id tag. If no matching constant is found, and empty hash is returned.
	def get_svninfo( classinfo )
		return {} unless classinfo[:sections]
		constants = classinfo[:sections].first[:constants] or return {}
	
		constants.find {|c| c[:value] =~ SVNID_PATTERN } or return {}

		filename, rev, date, time, committer = $~.captures
		commitdate = Time.parse( date + ' ' + time )
		
		return {
			:filename    => filename,
			:rev         => Integer( rev ),
			:commitdate  => commitdate,
			:commitdelta => time_delta_string( Time.now.to_i - commitdate.to_i ),
			:committer   => committer,
		}
	end


	### Load and render the erb template in the given +templatefile+ within the specified 
	### +context+ (a Binding object) and write it out to +outfile+. Both +templatefile+ and 
	### +outfile+ should be Pathname-like objects.
	def render_template( templatefile, context, outfile )
		template_src = templatefile.read
		template = ERB.new( template_src, nil, '<>' )
		template.filename = templatefile.to_s

		output = begin
			template.result( context )
		rescue NoMethodError => err
			raise "Error while evaluating %s: %s (at %p)" % [
				templatefile.to_s,
				err.message,
				eval( "_erbout[-50,50]", context )
			]
		end

		unless $dryrun
			outfile.dirname.mkpath
			outfile.open( 'w', 0644 ) do |ofh|
				ofh.print( output )
			end
		else
			debug_msg "  would have written %d bytes to %s" %
			[ output.length, outfile ]
		end
	end

end # Roc::Generator::Darkfish

# :stopdoc:

### Time constants
module TimeConstantMethods # :nodoc:
	
	### Number of seconds (returns receiver unmodified)
	def seconds
		return self
	end
	alias_method :second, :seconds

	### Returns number of seconds in <receiver> minutes
	def minutes
		return self * 60
	end
	alias_method :minute, :minutes  

	### Returns the number of seconds in <receiver> hours
	def hours
		return self * 60.minutes
	end
	alias_method :hour, :hours

	### Returns the number of seconds in <receiver> days
	def days
		return self * 24.hours
	end
	alias_method :day, :days

	### Return the number of seconds in <receiver> weeks
	def weeks
		return self * 7.days
	end
	alias_method :week, :weeks

	### Returns the number of seconds in <receiver> fortnights
	def fortnights
		return self * 2.weeks
	end
	alias_method :fortnight, :fortnights

	### Returns the number of seconds in <receiver> months (approximate)
	def months
		return self * 30.days
	end
	alias_method :month, :months

	### Returns the number of seconds in <receiver> years (approximate)
	def years
		return (self * 365.25.days).to_i
	end
	alias_method :year, :years


	### Returns the Time <receiver> number of seconds before the 
	### specified +time+. E.g., 2.hours.before( header.expiration )
	def before( time )
		return time - self
	end
	

	### Returns the Time <receiver> number of seconds ago. (e.g., 
	### expiration > 2.hours.ago )
	def ago
		return self.before( ::Time.now )
	end


	### Returns the Time <receiver> number of seconds after the given +time+.
	### E.g., 10.minutes.after( header.expiration )
	def after( time )
		return time + self
	end

	# Reads best without arguments:  10.minutes.from_now
	def from_now
		return self.after( ::Time.now )
	end
end # module TimeConstantMethods


# Extend Numeric with time constants
class Numeric # :nodoc:
	include TimeConstantMethods
end

